Analysis/src/EventCandidate.cpp

```diff
@@ -0,0 +1,138 @@
+    }
+
+    void EventCandidate::CreateLeptons()
+    {
+        double shifted_met_px = 0;
+        double shifted_met_py = 0;
+
+        if(uncertainty_source == UncertaintySource::TauES){
+          tuple_met = std::make_shared<ntuple::TupleMet>(event, MetType::PF);
+          met = std::shared_ptr<MET>(new MET(*tuple_met, tuple_met->cov()));
+        }
+
+
+        lepton_candidates = std::make_shared<LepCollection>();
+        tuple_leptons = std::make_shared<std::vector<ntuple::TupleLepton>>();
+        for(size_t n = 0; n < event.lep_p4.size(); ++n){
+          tuple_leptons->emplace_back(event, n);
+          ntuple::TupleLepton tuple_lepton = ntuple::TupleLepton(event, n);
+          lepton_candidates->push_back(LepCandidate(tuple_lepton,tuple_lepton.iso()));
+        }
+        for(size_t n = 0; n < tuple_leptons->size(); ++n) {
+          auto tuple_lepton = tuple_leptons->at(n);
+          analysis::LorentzVectorM lepton_p4(tuple_lepton.p4());
+          LorentzVectorM corrected_lepton_p4;
+          if(tuple_lepton.leg_type() == analysis::LegType::tau  && tuple_lepton.gen_match() == GenLeptonMatch::Tau){
+
+              UncertaintyScale current_scale = uncertainty_source == UncertaintySource::TauES ? scale : UncertaintyScale::Central;
+              double sf = GetCorrectionFactor(period,tuple_lepton.decayMode(),current_scale).first;
+
+
+              if(tuple_lepton.decayMode() == 0){
+                  double shifted_pt = lepton_p4.pt() * sf;
+                  corrected_lepton_p4 = LorentzVectorM(shifted_pt, lepton_p4.eta(), lepton_p4.phi(),lepton_p4.M());
+              }
+              else{
+                  corrected_lepton_p4 = lepton_p4 * sf;
+              }
+
+              shifted_met_px += tuple_lepton.p4().px() - corrected_lepton_p4.px();
+              shifted_met_py += tuple_lepton.p4().px() - corrected_lepton_p4.px();
+
+          }
+          lepton_candidates->at(n).SetMomentum(corrected_lepton_p4);
+        }
+
+        shifted_met_px += met->GetMomentum().px();
+        shifted_met_py += met->GetMomentum().py();
+        analysis::LorentzVectorXYZ shifted_met;
+        double E = std::hypot(shifted_met_px,shifted_met_py);
+        shifted_met.SetPxPyPzE(shifted_met_px,shifted_met_py,0,E);
+        met->SetMomentum(shifted_met);
+
+
+    }
+
+    void EventCandidate::CreateJets()
+    {
+        if(jecUncertainties->JetUncertainties_withTotal().count(uncertainty_source)){
+            tuple_met = std::make_shared<ntuple::TupleMet>(event, MetType::PF);
+            met = std::shared_ptr<MET>(new MET(*tuple_met, tuple_met->cov()));
+        }
+
+        jet_candidates = std::make_shared<JetCollection>();
+        tuple_jets = std::make_shared<std::vector<ntuple::TupleJet>>();
+        for(size_t n = 0; n < event.jets_p4.size(); ++n) {
+          tuple_jets->emplace_back(event, n);
+          ntuple::TupleJet tuple_jet = ntuple::TupleJet(event, n);
+          jet_candidates->push_back(JetCandidate(tuple_jet));
```

Comment:
it is important to use the same tuple jet as in tuple_jets and not the other one.